go/services/adr/client.go

```diff
@@ -0,0 +1,539 @@
+)
+
+// Error represents an error returned by the ADR service.
+type Error struct {
+	Message       string
+	PropertyName  string
+	PropertyValue string
+	Condition     interface{}
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+// Client manages interactions with the Azure Device Registry.
+type Client struct {
+	logger                                             *slog.Logger
+	protocol                                           *protocol.Application
+	mqtt                                               protocol.MqttClient
+	observedAeps                                       map[string]struct{}
+	observedAssets                                     map[string]struct{}
+	mu                                                 sync.RWMutex
+	closeOnce                                          sync.Once
+	listeners                                          protocol.Listeners
+	onAssetUpdate                                      func(aepName string, asset *Asset) error
+	onAepUpdate                                        func(aepName string, profile *AssetEndpointProfile) error
+	notifyOnAssetEndpointProfileUpdateInvoker          *adrbaseservice.NotifyOnAssetEndpointProfileUpdateCommandInvoker
+	notifyOnAssetUpdateInvoker                         *adrbaseservice.NotifyOnAssetUpdateCommandInvoker
+	getAssetEndpointProfileCommandInvoker              *adrbaseservice.GetAssetEndpointProfileCommandInvoker
+	updateAssetEndpointProfileStatusCommandInvoker     *adrbaseservice.UpdateAssetEndpointProfileStatusCommandInvoker
+	createDiscoveredAssetEndpointProfileCommandInvoker *aeptypeservice.CreateDiscoveredAssetEndpointProfileCommandInvoker
+	createDetectedAssetCommandInvoker                  *adrbaseservice.CreateDetectedAssetCommandInvoker
+	getAssetCommandInvoker                             *adrbaseservice.GetAssetCommandInvoker
+	updateAssetStatusCommandInvoker                    *adrbaseservice.UpdateAssetStatusCommandInvoker
+}
+
+// ClientOption allows setting optional parameters on the Client.
+type ClientOption func(*Client)
+
+// WithLogger sets a logger for the client.
+func WithLogger(logger *slog.Logger) ClientOption {
+	return func(c *Client) {
+		c.logger = logger
+	}
+}
+
+// WithAssetUpdateHandler sets a handler for asset update events.
+func WithAssetUpdateHandler(
+	handler func(aepName string, asset *Asset) error,
+) ClientOption {
+	return func(c *Client) {
+		c.onAssetUpdate = handler
+	}
+}
+
+// WithAepUpdateHandler sets a handler for asset endpoint profile update events.
+func WithAepUpdateHandler(
+	handler func(aepName string, profile *AssetEndpointProfile) error,
+) ClientOption {
+	return func(c *Client) {
+		c.onAepUpdate = handler
+	}
+}
+
+// New creates a new ADR client.
+func New(
+	app *protocol.Application,
+	client protocol.MqttClient,
+	opts ...ClientOption,
+) (*Client, error) {
+	notifyInvoker, err := adrbaseservice.NewNotifyOnAssetEndpointProfileUpdateCommandInvoker(
```

Comment:
Invokers are also listeners, so you probably want to include this in `c.listeners`. You also need to `Close()` everything if there's an error (which the `listeners` collection is also useful for). That being said, why aren't you using the generated wrapper(s) here, which should already be doing all that?